README.md
# 🆕 neu
## [pub.dev Listing](https://pub.dev/packages/neu) | [API Doc](https://pub.dev/documentation/neu/latest) | [GitHub](https://github.com/Zabadam/neu)
#### API References: [`Neu`](https://pub.dev/documentation/neu/latest/neu/Neu-class.html) | [`NeuContainer`](https://pub.dev/documentation/neu/latest/neu/NeuContainer-class.html) | [`Neu.toggle`](https://pub.dev/documentation/neu/latest/neu/Neu/Neu.toggle.html) | [`NeuToggle`](https://pub.dev/documentation/neu/latest/neu/NeuToggle-class.html) | [`Curvature`](https://pub.dev/documentation/neu/latest/neu/Curvature-class.html) | [`Swell`](https://pub.dev/documentation/neu/latest/neu/Swell-class.html) | [`NeuTextSpec`](https://pub.dev/documentation/neu/latest/neu/NeuTextSpec-class.html)
##### [🐸 Zaba.app ― simple packages, simple names.](#-zabaapp--simple-packages-simple-nameshttpspubdevpublisherszabaapppackages-other-flutter-packages-published-by-zabaapp)
## 🌐 [Interactive Web Demo](https://neu.zaba.app)

|   |   |
|:--|:--|
| [![neu icon](https://neu.zaba.app/img/icon.png)](https://pub.dev/packages/neu 'pub.dev listing') | A ***`Neu`morphic***, or *new [skeuomorphic](https://en.wikipedia.org/wiki/Skeuomorph)*, helper package and class with a variety of static and instance methods for the simple generation of shaded decorations conforming to the neumorphic design concept. |

<!-- [![neu header image](https://neu.zaba.app/img/header_300.gif)](https://pub.dev/packages/neu 'pub.dev listing') -->

These designs are often solid in terms of color variety: generally a
**dominant color** is used as a backdrop as well as the color of controls and
elements. These items would typically appear **on** or "above" the surface with
drop shadows in standard design systems but are intended in neumorphism to
extrude **from** or "rise out" or into the surface.

This extrusion, which can be imagined as *clay*- or *rubber*-like, is achieved
by a combination of same-color elements and background, contrasting shadows,
slight variances in color shading and (potentially) gradients for the impression
of **natural lighting**.

<br />

## 📚 Table of Contents

[![neu logo](https://neu.zaba.app/img/neu.png)](#neu)
  1. [Getting Started](#getting-started 'Getting Started')
  2. [Example](#example 'Example')
     1. [Depth](#depth 'Depth')
     2. [Spread](#spread 'Spread')
     3. [Curvature](#curvature 'Curvature')
     4. [Swell](#swell 'Swell')
     5. [Light Source](#light-source 'Light Source')
  3. [Generating Text Styles](#generating-text-styles 'Generating Text Styles')
  4. [Rapid Deploy with `NeuContainer`](#rapid-deploy-with-neucontainer 'Rapid Deploy with `NeuContainer`')
  5. [Permutations of `Curvature` X `Swell`](#permutations-of-curvature-x-swell 'Permutations of `Curvature` X `Swell`')
  6. 🛣️ [Roadmap](#️-roadmap 'The current TODO list')

<br />

## Getting Started

Create a `new Neu` object. This default constructor has all manner of defaults
in the cases where values are not provided.

```dart
const Neu({
  // Foundation
  Color color = lightWhite,
  int depth = defaultDepth, // 25
  double spread = defaultSpread, // 7.5

  // Fine-tuning
  Curvature curvature = Curvature.convex,
  Swell swell = Swell.emboss,
  Alignment lightSource = defaultLightSource, // Alignment.topLeft
  
  // For specific output method calls
  NeuTextSpec neuTextSpec = const NeuTextSpec(), // buildTextStyle()
  BorderRadius borderRadius = BorderRadius.zero, // buildBoxDecoration()
  ShapeBorder shape = defaultShape, // buildShapeDecoration()
})
```

This `Neu` object has obtainable properties that begin with "build". These
getters call `Neu`'s static methods with parameters filled from the object's
properties. 

Consider [`buildLinearGradient()`](https://pub.dev/documentation/neu/latest/neu/Neu/buildLinearGradient.html)
and [`buildBoxShadows()`](https://pub.dev/documentation/neu/latest/neu/Neu/buildBoxShadows.html);
combinations thereof: [`buildBoxDecoration()`](https://pub.dev/documentation/neu/latest/neu/Neu/buildBoxDecoration.html)
and [`buildShapeDecoration()`](https://pub.dev/documentation/neu/latest/neu/Neu/buildShapeDecoration.html);
or for typography, [`buildTextStyle()`](https://pub.dev/documentation/neu/latest/neu/Neu/buildTextStyle.html).

[![Demonstration of \`Neu.textStyle\`](https://neu.zaba.app/img/textStyle.png)](https://neu.zaba.app 'Goto web demo')

> `buildTextStyle()` with varying `depth` values
 
| ![transparent 600x1 pixel-pusher (ignore)](https://i.imgur.com/XCh0q2K.png) | [📚](#-table-of-contents 'Table of Contents') |
| --------------------------------------------------------------------------: | -------------------------------------------: |

As mentioned, the above properties call static helper methods by similar names.
These `Neu` static methods may be called at any time without instantiating an
object, like so:

```dart
final BoxDecoration neuDecoration = Neu.boxDecoration(
  color: Colors.indigoAccent.shade700,
  depth: 25,
  spread: 10,
  curvature: Curvature.flat,
  swell: Swell.emboss,
);
```
vs.
```dart
final BoxDecoration neuDecoration = Neu(
  color: Colors.indigoAccent.shade700,
  . . .
).buildBoxDecoration();
```

| ![transparent 600x1 pixel-pusher (ignore)](https://i.imgur.com/XCh0q2K.png) | [📚](#-table-of-contents 'Table of Contents') |
| --------------------------------------------------------------------------: | -------------------------------------------: |

## Example

Click any animation below to experience the 🌐 web demo, which is more or less
equivalent to the source code in the **[Example](https://pub.dev/packages/neu/example)**,
and is the source for the animations.

### Depth

With all `Neu` design the basis of these decorations is a `color` and a `depth`.
Color should ideally match or be similar to the color of the background behind
the resulting decoration. The `depth` is the **extent** to which this decoration
will appear *extruded* from its surface. A larger `depth` increases the contrast
of the shading of colors on either side of the decoration.

[![Demonstration of \`depth\`](https://neu.zaba.app/img/depth.gif)](https://neu.zaba.app 'Goto web demo')

| ![transparent 600x1 pixel-pusher (ignore)](https://i.imgur.com/XCh0q2K.png) | [📚](#-table-of-contents 'Table of Contents') |
| --------------------------------------------------------------------------: | -------------------------------------------: |

### Spread

In terms of shadows, the argument `spread` is responsible for determining how
wide an area the effect covers and how blurry the shadows appear.

Not all methods consider `spread`, notably `linearGradient()` (and
`buildLinearGradient()` by extension) create a decoration―`LinearGradient`―that
does not have shadows. These gradients may be combined with shadows in other
situations, where `spread` will then be considered.

[![Demonstration of \`spread\`](https://neu.zaba.app/img/spread.gif)](https://neu.zaba.app 'Goto web demo')

| ![transparent 600x1 pixel-pusher (ignore)](https://i.imgur.com/XCh0q2K.png) | [📚](#-table-of-contents 'Table of Contents') |
| --------------------------------------------------------------------------: | -------------------------------------------: |

### Curvature

A `Curvature` is a description of the appearance of the actual surface of
the decoration. A `Curvature.flat` decoration has no gradient (or rather, a
solid-color `Gradient`), while `Curvature.convex` orders a light -> dark
gradient in a way that inspires a "bubble"-like appearance.

Other options make the effect more extreme, such as `Curvature.superconvex`,
or reverse the effect: `Curvature.concave`.

See the [twenty permutations of `Curvature` X `Swell`](#permutations-of-curvature-x-swell).

[![Demonstration of \`curvature\`](https://neu.zaba.app/img/curvature.gif)](https://neu.zaba.app 'Goto web demo')

| ![transparent 600x1 pixel-pusher (ignore)](https://i.imgur.com/XCh0q2K.png) | [📚](#-table-of-contents 'Table of Contents') |
| --------------------------------------------------------------------------: | -------------------------------------------: |

### Swell

A `Swell` is an overall depiction of how the decoration appears in terms of
being inset into the surface or extruded from it.

- Combining [Swell.emboss] with [Curvature.convex], the decoration will appear
like a bubble lifting out of the background.

- Combining [Swell.deboss] with [Curvature.concave], the decoration will appear
like a bubble popped and depressed into the background.

This property, especially [when combined with varying `Curvature`s](#permutations-of-curvature-x-swell)
can really "sell" the *pressed* or *unpressed* effect.

[![Demonstration of \`swell\`](https://neu.zaba.app/img/swell.gif)](https://neu.zaba.app 'Goto web demo')

| ![transparent 600x1 pixel-pusher (ignore)](https://i.imgur.com/XCh0q2K.png) | [📚](#-table-of-contents 'Table of Contents') |
| --------------------------------------------------------------------------: | -------------------------------------------: |

### Light Source

The `lightSource` is always set by default as `defaultLightSource`, which is
`Alignment.topLeft`. This gives the illusion of lighting the entire neumorphic
decoration from the top-left corner. All descriptions of gradient and shadow
directionality and the illusion of being *toggled* or *not toggled* are based
on this default light source. An overriding `Alignment` may be provided,
however, to dynamically "relight" the decorations.

- Consider this lighting entirely *artificial*. Aspects of real light physics
  are not recreated. Simply put, this value is used to offset/shift the light
  and dark shadows.

[![Demonstration of \`lightSource\`](https://neu.zaba.app/img/lightSource.gif)](https://neu.zaba.app 'Goto web demo')

| ![transparent 600x1 pixel-pusher (ignore)](https://i.imgur.com/XCh0q2K.png) | [📚](#-table-of-contents 'Table of Contents') |
| --------------------------------------------------------------------------: | -------------------------------------------: |

## Generating Text Styles

When using a `Neu` object and its `buildTextStyle()` property to generate
neumorphic text decorations as opposed to the static method `Neu.textStyle()`,
then instantiate and customize a [`NeuTextSpec` object](https://pub.dev/documentation/neu/latest/neu/NeuTextSpec-class.html)
as well. This secondary parameter container object cleans up the `Neu`
constructor's arguments.

> Other secondary constructor arguments `borderRadius` and `shape` apply to
> instance methods [`buildBoxDecoration()`](https://pub.dev/documentation/neu/latest/neu/Neu/buildBoxDecoration.html)
> and [`buildShapeDecoration()`](https://pub.dev/documentation/neu/latest/neu/Neu/buildShapeDecoration.html)
> respectively.

| ![transparent 600x1 pixel-pusher (ignore)](https://i.imgur.com/XCh0q2K.png) | [📚](#-table-of-contents 'Table of Contents') |
| --------------------------------------------------------------------------: | -------------------------------------------: |

## Rapid Deploy with `NeuContainer`

In some circumstances one may want to place a neumorphic element that does not
already have a solid-matching-color background; or perhaps one would simply
prefer to deploy *`Neu`morphic* decorations with ease and animation support;
in either situation a [`NeuContainer`](https://pub.dev/documentation/neu/latest/neu/NeuContainer/NeuContainer.html)
may come in handy.

This glorified two-layer `AnimatedContainer` has a special `EdgeInsets` property
called [`insets`](https://pub.dev/documentation/neu/latest/neu/NeuContainer/insets.html)
that adds additional padding beyond `padding` with the same color as `color`.

| ![transparent 600x1 pixel-pusher (ignore)](https://i.imgur.com/XCh0q2K.png) | [📚](#-table-of-contents 'Table of Contents') |
| --------------------------------------------------------------------------: | -------------------------------------------: |

##  Permutations of `Curvature` X `Swell`

[![Twenty permutations of \`curvature\` X \`swell\`](https://neu.zaba.app/img/permutations.png)](https://neu.zaba.app 'Goto web demo, swipe to Page 5')

> This is the fifth page (by swiping) in the **Example** app.

| ![transparent 600x1 pixel-pusher (ignore)](https://i.imgur.com/XCh0q2K.png) | [📚](#-table-of-contents 'Table of Contents') |
| --------------------------------------------------------------------------: | -------------------------------------------: |

## 🛣️ Roadmap

- [ ] Bespoke `NeuFoo` pre-built `Widget`s

> Imagine a `NeuAppBar` and tab bar, `NeuScrollbar` or `NeuSlider`.
> 
> At the moment consider `Neu.toggle` named constructor which simplifies the
> customization of neumorphic appearance down to a few `bool` flags. These toggles
> mesh well with the existing paradigm for buttons to appear *pressed* or
> *not pressed*.
> 
> Furthermore [`new NeuToggle`](https://pub.dev/documentation/neu/latest/neu/NeuToggle-class.html)
> may be appropriate for some situations as a `StatefulWidget` that manages its
> own `Neu.toggle` and `GestureDetector`, rendering like an `AnimatedContainer`.
> 
> | Large Widget | Small Widget |
> |:---------------------:|:----------------------:|
> | [![Demonstration of \`NeuToggle\`, large widget](https://neu.zaba.app/img/toggle_big.gif)](https://neu.zaba.app 'Goto web demo') | [![Demonstration of \`NeuToggle\`, small widget](https://neu.zaba.app/img/toggle_small.gif)](https://neu.zaba.app 'Goto web demo') |
> |  |  |
> 
> Each `NeuToggle` includes a `providesFeedback` flag as a convenience shortcut
> for `HapticFeedback.vibrate()` which is `false` by default. 
> 
> Pressing-and-holding (versus a single tap) will call the `onToggle` callback,
> in `void Function(bool isToggled)` format, twice: once when the long-press is
> triggered and again when the long-press is released. The status of the toggle,
> with default normal value of `false`, is the boolean delivered with the
> callback.

- [ ] Debug Props, hash, equality checks, etc.

- [ ] Demo / Example app

> Is already as good as it needs to be, but could be beefed up just a tad to make
> it handy, say, as a generator for decoration code without needing the package.

- [ ] Integration with [`surface`](https://pub.dev/packages/surface)

- [ ] Integration with sensors data for responsive light source

> Planned integration with upcoming package [`sense`](https://pub.dev/packages/sense)
> (around the time existing packages [`xl`](https://pub.dev/packages/xl) and
> [`foil`](https://pub.dev/packages/foil) are transitioned to it).

- [ ] Inner `Shadow`s (may just be achieved through [`surface`](https://pub.dev/packages/surface))

- [ ] Custom `Neu` Painter

- [ ] Interpolation

> Already works pretty well through the nature of these decorations, but bespoke
> tweenage and lerping could be written if it seems helpful at some point.

| ![transparent 600x1 pixel-pusher (ignore)](https://i.imgur.com/XCh0q2K.png) | [📚](#-table-of-contents 'Table of Contents') |
| --------------------------------------------------------------------------: | -------------------------------------------: |

<br />

---

## 🐸 [Zaba.app ― simple packages, simple names.](https://pub.dev/publishers/zaba.app/packages 'Other Flutter packages published by Zaba.app')

<details>
<summary>More by Zaba</summary>

### Wrappers | Widgets that surround other widgets with functionality
- ## 🕹️ [xl](https://pub.dev/packages/xl 'implement accelerometer-fueled interactions with a layering paradigm')
- ## 🌈 [foil](https://pub.dev/packages/foil 'implement accelerometer-reactive gradients in a cinch')
- ## 📜 [curtains](https://pub.dev/packages/curtains 'provide animated shadow decorations for a scrollable to allude to unrevealed content')

---
### Container Widget | Wraps many functionalities in one, very customizable
- ## 🌟 [surface](https://pub.dev/packages/surface 'animated, morphing container with specs for Shape, Appearance, Filter, Tactility')

---
### Succinct Utility | Work great alone or employed above
- ## 🆕 [![neu header](https://neu.zaba.app/img/neu_20.png)](https://pub.dev/packages/neu 'A helper for generating outstanding neumorphic-conforming designs')
- ## 🙋‍♂️ [img](https://pub.dev/packages/img 'An extended Image \"Too\" and DecorationImageToo that support an expanded Repeat.mirror painting mode')
- ## 🙋‍♂️ [icon](https://pub.dev/packages/icon 'An extended Icon \"Too\" for those that are not actually square, plus shadows support + IconUtils')
- ## 🏓 [ball](https://pub.dev/packages/ball 'A bouncy, position-mirroring splash factory that\'s totally customizable')
- ## 👁‍🗨 [sense](https://pub.dev/packages/sense 'A widget that outputs actionable sensors data with intervals, delays, normalization, and more.')
- ## 👥 [shadows](https://pub.dev/packages/shadows 'Convert a double-based \`elevation\` + BoxShadow and List\<BoxShadow\> extensions')
- ## 🎨 [![spectrum header](https://raw.githubusercontent.com/Zabadam/spectrum/main/doc/img/spectrum_15.gif)](https://pub.dev/packages/spectrum 'Color and Gradient utilities such as GradientTween, copyWith, complementary for colors, AnimatedGradient, MaterialColor generation, and more')
</details>


CHANGELOG.md
# neu
## **[0.1.1]** - 25 AUG 21
- Adds proper README.
- Fixes bug: no passage of `shape` to lower, secondary `insets`-provided
  container when building `NeuContainer` and `NeuToggle`.

## **[0.1.0]** - 21 AUG 21
- Initial release.


lib/src/widgets/container.dart
/// Provides [NeuContainer], a specialized [StatelessWidget] that builds two
/// layered [AnimatedContainer]s, one as a solid color backdrop with edge insets
/// and the second conforming to the [Neu]-standard properties to give the inner
/// container a neumorphic or "clay"-like appearance rising out of the lower
/// container.
library neu;

import '../common.dart';
import '../neu.dart';
import 'toggle.dart';

/// Construct a [new NeuContainer], a specialized [StatelessWidget] that builds
/// two layered [AnimatedContainer]s, one as a solid color backdrop with edge
/// [insets] and the second conforming to the [Neu]-standard properties, such as
/// [depth] and [curvature], to give the inner container a neumorphic or
/// "clay"-like appearance rising out of the lower container.
class NeuContainer extends StatelessWidget {
  /// A specialized [StatelessWidget] that builds two layered
  /// [AnimatedContainer]s, one as a solid color backdrop with edge [insets]
  /// and the second conforming to the [Neu]-standard properties, such as
  /// [depth] and [curvature], to give the inner container a neumorphic or
  /// "clay"-like appearance rising out of the lower container.
  ///
  /// ---
  /// #### SEE ALSO:
  ///
  /// * [Neu], the base class defining this design system.
  /// * [AnimatedContainer], which this `Widget` will render and from which
  ///   many if its parameters are copied.
  /// * [NeuToggle] for a widget like this container but that handles its own
  ///   [GestureDetector] for the purpose of "toggling" between two
  ///   *[Neu]morphic* states.
  const NeuContainer({
    Key? key,
    this.color = lightWhite,
    this.depth = defaultDepth,
    this.insets = EdgeInsets.zero,
    this.curvature = Curvature.convex,
    this.swell = Swell.emboss,
    this.spread = defaultSpread,
    this.lightSource = defaultLightSource,

    /// Container
    this.alignment,
    this.width,
    this.height,
    this.margin,
    this.padding,
    this.constraints,
    this.shape = defaultShape,
    this.foregroundDecoration,
    this.transform,
    this.transformAlignment,
    this.duration = defaultDuration,
    this.curve = Curves.elasticOut,
    this.onEnd,
    this.child,
  }) : super(key: key);

  /// With all [Neu] design the basis of these decorations is a [color] and a
  /// [depth]. Color should ideally match with or be similar to the color of the
  /// background behind the resulting decoration. The `depth` is the extent to
  /// which this decoration will appear "extruded" from its surface. A larger
  /// `depth` increases the contrast of the shading of colors on either side of
  /// the decoration.
  ///
  /// Default is [lightWhite], `Color(0xFFE0E0E0)`.
  ///
  /// If [insets] is given a non-zero value, this [color] will also be used as
  /// the background for a secondary container under the neumorphic one,
  /// providing a padded, solid-color backdrop for the container.
  final Color color;

  /// With all [Neu] methods the basis of these decorations is a [color] and a
  /// [depth]. Color should ideally match or be similar to the color of the
  /// background behind the resulting decoration. The `depth` is the extent to
  /// which this decoration will appear "extruded" from its surface. A larger
  /// `depth` increases the contrast of the shading of colors on either side of
  /// the decoration.
  ///
  /// Default is [defaultDepth], `25`.
  final int depth;

  /// A convenience to provide a padded, solid-color backdrop on which the true
  /// container will be rendered. In a scenario where a background is not a
  /// solid color that matches [color], then [insets] may be initialized
  /// non-negligibly (such as `EdgeInsets.all(25)`) to provide a small platform
  /// for the *[Neu]morphic* effect to be more visible.
  final EdgeInsetsGeometry insets;

  /// A [Curvature] is a description of the appearance of the actual surface of
  /// the decoration. A [Curvature.flat] decoration has no gradient (solid-color
  /// [Gradient]), while [Curvature.convex] orders a light -> dark gradient in a
  /// way that inspires a "bubble"-like appearance, especially when combined
  /// with a [swell] of [Swell.emboss].
  /// - Conversely, to pop that bubble like its bubble-wrap, [Swell.deboss]
  ///   can be combined with [Curvature.concave].
  ///
  /// *(Other options make the effect more extreme, namely
  /// [Curvature.superconvex] and [Curvature.superconcave].)*
  ///
  /// Each constant in the [Curvature] enum has a description of how it affects
  /// various areas in its usage.
  ///
  /// The default is [Curvature.flat], the result from which will be a
  /// solid-color (non-shaded) gradient.
  final Curvature curvature;

  /// A [Swell] is an overall depiction of how the decoration appears in terms
  /// of being inset into the surface or extruded from it.
  ///
  /// Each constant in the [Swell] enum has a description of how it affects
  /// various areas in its usage.
  ///
  /// The default is [Swell.emboss] which make a visual foundation for an
  /// element to be clickable or *not toggle* in a toggleable scenario.
  /// Combines great with [Curvature.flat] or [Curvature.convex] for a
  /// bubble-like appearance.
  ///
  /// The opposite would be [Swell.deboss]. This state gives a good impression
  /// (😏) that an element has been clicked or *toggled on* in a toggleable
  /// scenario. Combines great with [Curvature.concave] for an inset, popped
  /// bubble-wrap appearance.
  ///
  /// The effect is made more dramatic by [Swell.superemboss] or
  /// [Swell.superdeboss].
  final Swell swell;

  /// In terms of [Shadow]s, [spread] is responsible for determining how wide
  /// an area the effect covers and how blurry the shadows appear.
  ///
  /// The default is [defaultSpread], `7.5`.
  final double spread;

  /// The [lightSource] is always set by default as [defaultLightSource], which
  /// is [Alignment.topLeft]. This gives the illusion of lighting the entire
  /// neumorphic decoration from the top-left corner. All descriptions of
  /// gradient and shadow directionality and the illusion of being *toggled*
  /// or *not toggled* are based on this default light source. An overriding
  /// [Alignment] may be provided, however, to dynamically "relight" the
  /// decorations.
  /// - Consider this *lighting* entirely artificial. Aspects of real light
  ///   physics are not recreated. Simply put, this value is used to
  ///   offset/shift the light and dark shadows.
  ///
  /// This value is also used as this container's [LinearGradient.begin]
  /// property.
  final Alignment lightSource;

  /// A [ShapeBorder] object provides a description of a shape for a decoration.
  /// It can include information such as how round corners are to be or what the
  /// border edges should look like.
  ///
  /// Default is [defaultShape], a [RoundedRectangleBorder] with a
  /// [defaultRadius] `borderRadius` of `5.0` and no border.
  final ShapeBorder shape;

  // --- Container --- //

  /// Limit the [width] or [height] of this [NeuContainer], excluding any
  /// [margin], in logical pixels.
  ///
  /// Any defined [insets] and [padding] will fit within this range. Outside of
  /// the dimensions defined by these optional fields lies any [EdgeInsets]
  /// from [margin].
  final double? width, height;

  /// Empty space to surround the [NeuContainer] and any [insets].
  ///
  /// If this [NeuContainer] is constrained or has a supplied [width]/[height],
  /// that dimensional description fits within these [EdgeInsets].
  ///
  /// That is to say that the net size of this `Widget` will be any constrained
  /// size (or provided by [child], if non-`null`) *plus* the insets provided
  /// here.
  final EdgeInsetsGeometry? margin;

  /// Empty space to inscribe inside the [Container]. The [child], if any, is
  /// placed inside these [EdgeInsets].
  final EdgeInsetsGeometry? padding;

  /// Additional constraints to apply to the child.
  ///
  /// The [width] and [height] fields are combined with these [constraints].
  ///
  /// The [insets] then [padding] then [child] go inside these constraints.
  final BoxConstraints? constraints;

  /// Align the [child] within the container.
  ///
  /// If non-`null`, the container will expand to fill its parent and position
  /// its child within itself according to the given value. If the incoming
  /// constraints are unbounded, then the child will be shrink-wrapped instead.
  ///
  /// Ignored if [child] is `null`.
  ///
  /// See also:
  ///
  ///  * [Alignment], a class with convenient constants typically used to
  ///    specify an [AlignmentGeometry].
  ///  * [AlignmentDirectional], like [Alignment] but for specifying alignments
  ///    relative to text direction.
  final AlignmentGeometry? alignment;

  /// The [Decoration] to paint in *front* of the [child].
  final Decoration? foregroundDecoration;

  /// The transformation matrix to apply before painting the container.
  ///
  /// See [Matrix4].
  final Matrix4? transform;

  /// The alignment of the origin, relative to the size of the container, if
  /// [transform] is specified.
  ///
  /// When [transform] is null, the value of this property is ignored.
  ///
  /// See also:
  ///
  ///  * [Transform.alignment], which is set by this property.
  final AlignmentGeometry? transformAlignment;

  /// How long any changes to the visual properties of this [NeuContainer] will
  /// take to fully propagate and animate. Changes are alterations to the
  /// properties/fields of this [NeuContainer], such as [padding] or [color].
  ///
  /// Default is [defaultDuration] (`1200ms`) which, when paired with a springy
  /// [curve] such as the default [Curves.elasticOut], results in a delicious
  /// bouncing neumorphic transition.
  final Duration duration;

  /// The [Curve] by which to implicitly animate any changes to the properties
  /// of this [NeuContainer]. Specifically, altering one of the fields, such as
  /// modifying [padding] or [color], will animate changes over this animation
  /// curve.
  ///
  /// Default is [Curves.elasticOut] which, when paired with a longer [duration]
  /// such as the default [defaultDuration] (`1200ms`), results in a delicious
  /// bouncing neumorphic transition.
  final Curve curve;

  /// An optional function to perform any time the properties of this
  /// [NeuContainer] are altered.
  final VoidCallback? onEnd;

  /// The [child] contained by the container.
  ///
  /// If null, and if the [constraints] are unbounded or also null, the
  /// container will expand to fill all available space in its parent, unless
  /// the parent provides unbounded constraints, in which case the container
  /// will attempt to be as small as possible.
  ///
  /// {@macro flutter.widgets.ProxyWidget.child}
  final Widget? child;

  @override
  Widget build(BuildContext context) => AnimatedContainer(
        width: width,
        height: height,
        constraints: constraints,
        margin: margin,
        padding: insets,
        decoration: ShapeDecoration(
          shape: shape,
          color: insets == EdgeInsets.zero ? Colors.transparent : color,
        ),
        duration: duration,
        curve: curve,
        child: AnimatedContainer(
          decoration: Neu(
            color: color,
            curvature: curvature,
            swell: swell,
            depth: depth,
            spread: spread,
            lightSource: lightSource,
            shape: shape,
          ).buildShapeDecoration,
          alignment: alignment,
          padding: padding,
          foregroundDecoration: foregroundDecoration,
          transform: transform,
          transformAlignment: transformAlignment,
          duration: duration,
          curve: curve,
          onEnd: onEnd, // Should call in parent?
          child: child,
        ),
      );

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties
      ..add(ColorProperty('color', color))
      ..add(IntProperty('depth', depth))
      ..add(DiagnosticsProperty<Curvature>('curvature', curvature))
      ..add(DiagnosticsProperty<Swell>('swell', swell))
      ..add(DoubleProperty('spread', spread))
      ..add(DiagnosticsProperty<Alignment>('lightSource', lightSource))
      ..add(DiagnosticsProperty<EdgeInsetsGeometry>('insets', insets,
          defaultValue: null))
      ..add(DiagnosticsProperty<ShapeBorder>('shape', shape))
      ..add(DoubleProperty('width', width, defaultValue: null))
      ..add(DoubleProperty('height', height, defaultValue: null))
      ..add(DiagnosticsProperty<EdgeInsetsGeometry>('margin', margin,
          defaultValue: null))
      ..add(DiagnosticsProperty<EdgeInsetsGeometry>('padding', padding,
          defaultValue: null))
      ..add(DiagnosticsProperty<BoxConstraints>('constraints', constraints,
          defaultValue: null))
      ..add(DiagnosticsProperty<AlignmentGeometry>('alignment', alignment,
          showName: false, defaultValue: null))
      ..add(DiagnosticsProperty<Decoration>(
          'foregroundDeco', foregroundDecoration,
          defaultValue: null))
      ..add(ObjectFlagProperty<Matrix4>.has('transform', transform))
      ..add(DiagnosticsProperty<AlignmentGeometry>(
          'transformAlignment', transformAlignment,
          defaultValue: null));
  }
}
